Mathlib/Tactic/Linter/DocString.lean

```diff
@@ -0,0 +1,86 @@
+/-
+Copyright (c) 2025 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Rothgang, Damiano Testa
+-/
+
+import Mathlib.Tactic.Linter.Header
+
+/-!
+#  The "DocString" style linter
+
+The "DocString" linter validates style conventions regarding doc-string formatting.
+-/
+
+open Lean Elab
+
+namespace Mathlib.Linter
+
+/--
+The "DocString" linter validates style conventions regarding doc-string formatting.
+-/
+register_option linter.style.docString : Bool := {
+  defValue := false
+  descr := "enable the style.docString linter"
+}
+
+namespace Style
+
+/-- Checks whether a declaration docstring `input` conforms to mathlib's style guidelines
+(or, at least the syntactically checkable parts).
+If the doc-string is not well-formed, return `some messages` where `messages` describe
+what went wrong, otherwise return `none`. -/
+def checkDocstring (initialWhitespace input : String) : Option (Array String) := do
+  let mut errors := #[]
+  match initialWhitespace with
+  | "\n" | " " => pure ()
+  | "" =>
+    errors := errors.push s!"error: doc-strings should start with a space or newline"
+  | _ =>
+    -- In any other cases, we have extraneous whitespace.
+    errors := errors.push s!"error: doc-strings should start with a single space"
+
+  -- Check the ending of the doc-string: a new line or exactly one space.
+  if !(input.endsWith "\n" || input.endsWith " ") then
+    errors := errors.push s!"error: doc-strings should end with a space or newline"
+  else if (input.endsWith "  ") then
+    errors := errors.push s!"error: doc-strings should end with at most a single space"
+  -- Catch misleading indentation.
+  let lines := (input.split (· == '\n')).drop 0
+  if lines.any (·.startsWith " ")
+    -- For now, we skip cases where a line starts with "* " or "- " as these are probably markdown
+    -- code blocks. (We can later try to re-enable some linting there.)
+    && !lines.any (fun l ↦ l.startsWith "* " || l.startsWith "- ") then
```

Comment:
How about including also ```` ``` ```` as an allowed start for exceptions?

Overall, I am still not sold on this check, as I find that it would require a more thorough parsing of markdown to really be effective.